### PR TITLE
Fix tracking lost ground contact

### PR DESCRIPTION
**Pull request by julianschuler**
## Introduced Changes

This PR fixes tracking lost ground contact during penalized.

## ToDo / Known Issues

- [x] Test on a field (currently all occupied)

## Ideas for Next Iterations (Not This PR)

None.

## How to Test

Upload to a robot and have fun penalizing and unpenalizing while picking up the robot, moving it around, etc.

### CODE REVIEW

File: crates/control/src/localization.rs [hunk 1]
```diff
@@ -208,26 +208,44 @@ impl Localization {
...
-                    }) || !context.has_ground_contact) =>
-            {
-                if self.is_penalized_with_motion_in_set_or_initial {
-                    if self.was_picked_up_while_penalized_with_motion_in_set_or_initial {
-                        self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
-
+                    })
+                {
+                    if self.is_penalized_with_motion_in_set_or_initial {
+                        if self.was_picked_up_while_penalized {
```
Comment (schluis) [inline crates/control/src/localization.rs:223]: Please restructure this so you don't have to check it twice. I think this is quite confusing to read.

File: crates/control/src/localization.rs [hunk 2]
```diff
@@ -208,26 +208,44 @@ impl Localization {
...
+                                .into_iter()
+                                .map(|pose| {
+                                    ScoredPose::from_isometry(
+                                        pose,
+                                        *context.penalized_hypothesis_covariance,
+                                        *context.initial_hypothesis_score,
+                                    )
+                                })
+                                .collect();
+                        }
```
Comment (schluis) [inline crates/control/src/localization.rs:240]: Why do you regenerate this and refill this?  Would it be sufficient to copy the `hypotheses_when_entered_playing`?

File: crates/control/src/localization.rs [hunk 3]
```diff
@@ -208,26 +208,44 @@ impl Localization {
...
-            {
-                if self.is_penalized_with_motion_in_set_or_initial {
-                    if self.was_picked_up_while_penalized_with_motion_in_set_or_initial {
-                        self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
-
+                    })
+                {
+                    if self.is_penalized_with_motion_in_set_or_initial {
+                        if self.was_picked_up_while_penalized {
+                            self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
```
Comment (schluis) [inline crates/control/src/localization.rs:224]: I think we should do something more intelligent: sometimes, the robot is picked up by accident during `is_penalized_with_motion_in_set_or_initial` and is placed at a similar location as before. We could keep the old hypothesis, reduce their probability and add the default penalized poses

**Review by schluis — CHANGES_REQUESTED** [review]

File: crates/control/src/localization.rs (hunk 1 above)
Comment (julianschuler) [reply crates/control/src/localization.rs]: Thanks, done.

File: crates/control/src/localization.rs (hunk 2 above)
Comment (julianschuler) [reply crates/control/src/localization.rs]: Good catch, done.

File: crates/control/src/localization.rs (hunk 3 above)
Comment (julianschuler) [reply crates/control/src/localization.rs]: I agree, though this should be done in a separate PR. Feel free to open an issue for that.

File: crates/control/src/localization.rs (hunk 1 above)
Comment (schluis) [reply crates/control/src/localization.rs]: Doesn't seem like it, even after switching to the correct branch 😉

Comment (julianschuler) [reply crates/control/src/localization.rs]: Could you please elaborate what you mean? The previously included double check of `self.was_picked_up_while_penalized` is not present anymore.

**Review by schluis — APPROVED** [review]

**Pull request merged**